### ISSUE TITLE
Use lazy ingestion in POMAnalyzer for newly discovered dependencies

### ISSUE THREAD

**Issue by mir-am** [fasten-project/fasten#352]
Currently, POMAnalyzer inserts newly discovered dependencies for Maven packages as a package into the metadata DB.
See:
https://github.com/fasten-project/fasten/blob/03c4ef0bf9924af2f4d25183a61a9cfc96a29286/analyzer/pom-analyzer/src/main/java/eu/fasten/analyzer/pomanalyzer/POMAnalyzerPlugin.java#L258

We can insert this as a package version and also use a REST API call to lazy ingest it so that we will have CGs and GIDs.

**Comment by MagielBruntink** [comment]
@mir-am I gave the processing of dependencies some further thought. If we have the POM Analyzer request ingestion dependencies through the REST API, pretty soon we will have a massive number of package-version to processes. Because it will request dependencies of dependencies, etc. and the REST API will put them all on the priority topics. I think it will be better to do this on the normal topic, so directly from the POM Analyzer.

**Comment by mir-am** [comment]
Good point!

**Comment by proksch** [comment]
I realized that the current behavior is likely the explanation for the problem raised in #347 and I am afraid that I can only think about a different solution...

The problem is simple. If the dependencies of an artifact will only be processed *after* the artifact itself, then we have problems further down in the pipeline: when we try to stitch the merged callgraph for said artifact to create the callable index, we can only consider those dependencies that are already in the database, i.e., those that have already been required by other artifacts earlier. Unknown dependencies will simply not be part of the CG merging. This is a problem!

I can only think about one single solution to this: when ingesting an artifact, we first need to check for all (direct) dependencies whether they have been ingested already and add those that have not been before processing the requested artifact. This processing needs to be recursive, as the same problem exists for direct dependencies as well (we first need to ingest their dependencies and so on). Simply resolving the dependency set does not work, as order is not considered there.

This also means that the dependencies need to be added on the same "lane" (norm/prio) than the requested artifact, as we must enforce processing order. This will become faster, the more complete the database is, but in the beginning this will likely introduce a bit of lag for lazy ingested artifacts.

To keep things simple and not duplicate logic between the crawler and the lazy ingestion via REST, I am wondering, whether we should actually just implement this logic in the POMAnalyzer, as this seems to be the first plugin that actually knows about dependencies. On top of what I have described before, all that is required there is to track these "shortcuts" to avoid that they will be ingested again at some point, but this only means adding them to the correct table. Also, we need to prevent timeouts for the Kafka consumers, but [we know how to do this](https://github.com/fasten-project/fasten/pull/339) now. :)

**Comment by MagielBruntink** [comment]
> I realized that the current behavior is likely the explanation for the problem raised in #347 and I am afraid that I can only think about a different solution...
> 
> The problem is simple. If the dependencies of an artifact will only be processed _after_ the artifact itself, then we have problems further down in the pipeline: when we try to stitch the merged callgraph for said artifact to create the callable index, we can only consider those dependencies that are already in the database, i.e., those that have already been required by other artifacts earlier. Unknown dependencies will simply not be part of the CG merging. This is a problem!

I was not aware that the callable indexer tries to merge callgraphs. Does it do that for every package-version entering the pipeline?

> I can only think about one single solution to this: when ingesting an artifact, we first need to check for all (direct) dependencies whether they have been ingested already and add those that have not been, before processing the requested artifact. This processing needs to be recursive, as the same problem exists for direct dependencies as well (we first need to ingest their dependencies and so on). Simply resolving the dependency set does not work, as order is not considered there.
> 
> This also means that the dependencies need to be added on the same "lane" (norm/prio) than the requested artifact, as we must enforce processing order. This will become faster, the more complete the database is, but in the beginning this will likely introduce a bit of lag for lazy ingested artifacts.

This will undoubtedly lead to massive balls of dependencies sitting in the queue.

> To keep things simple and not duplicate logic between the crawler and the lazy ingestion via REST, I am wondering, whether we should actually just implement this logic in the POMAnalyzer, as this seems to be the first plugin that actually knows about dependencies. On top of what I have described before, all that is required there is to track these "shortcuts" to avoid that they will be ingested again at some point, but this only means adding them to the correct table. Also, we need to prevent timeouts for the Kafka consumers, but [we know how to do this](https://github.com/fasten-project/fasten/pull/339) now. :)

The downside of changing POM Analyzer will be that the issues is fixed only for Maven. What about Debian and PyPI? 

Would it be possible to address the issue later in the pipeline, eg. at the callable indexer itself? Would it be possible to incrementally update the callable index, if new dependencies for a package-version are coming in?

**Comment by mir-am** [comment]
@proksch, @MagielBruntink
I also think that issue #347 will be solved by processing direct dependencies recursively in the pipeline.
The current solution that I can come up ATM:
1- Let POMAnalyzer process Maven coordinates and get their direct dependencies.
2- When adding direct deps. to the DB, invoke a REST API call to lazy-ingest them.
3- Direct deps. will recursively be added to `fasten.mvn.priority.ingest` and will be processed further down the pipeline, i.e. OPAL, etc.

The possible downsides of the above approach is that:
1- Tons of packages will be added to the priority topic of POMAnalyzer.
2- There will be repetitive computation for dependencies that are already processed. This can be solved by checking whether a package version is already inserted into the DB.

**Comment by MagielBruntink** [comment]
> 2- There will be repetitive computation for dependencies that are already processed. This can be solved by checking whether a package version is already inserted into the DB.
I think the REST API does that - if a package-version is already in the DB (package_versions table) or already requested for ingestion (ingested_artifacts table), it will not fire a new request.

**Comment by mir-am** [comment]
> > 2- There will be repetitive computation for dependencies that are already processed. This can be solved by checking whether a package version is already inserted into the DB.
> 
> I think the REST API does that - if a package-version is already in the DB (package_versions table) or already requested for ingestion (ingested_artifacts table), it will not fire a new request.

Yes! The REST API does check whether the package version exists.
If all agree on this solution, I can implement it in POMAnalyzer and create a PR.

**Comment by MagielBruntink** [comment]
I don't know if it's the right solution, and it will come at a large cost of delayed processing of coordinates.

Open questions in my view:
 - what does the callable indexer actually try to accomplish, why is it merging graphs?
 - how will the callable indexer decide that everything it needs is present? Will it assume that if it gets a message for a package-version, all the work on its dependencies has been completed?
 - what if some dependencies fail to process (and they often do), what will the callable indexer do then?
 - why not incrementally update the graph DB of the callable indexer for each new package-version that it processes. I don't understand why that would not be possible.
 - let's hope there are no cycles in the Maven dependency graph! Did anybody every check? Or whether our resolution algorithm can create dependency cycles.

**Comment by proksch** [comment]
> I don't know if it's the right solution, and it will come at a large cost of delayed processing of coordinates.

The problem is that this is the nature of the beast. To get a full call graph, all dependencies need to be in the KB. The more dependencies are missing, the less precise the CG.

> what does the callable indexer actually try to accomplish, why is it merging graphs? how will the callable indexer decide that everything it needs is present? Will it assume that if it gets a message for a package-version, all the work on its dependencies has been completed?

The callable indexer uses this to index all source/target pairs after stitching, which is an information that cannot be stored in the KB. In the proposed solution, the callable indexer does not need to decide anything. The assumption is indeed that by adding the packages in the right order into the pipeline, everything will be available in time.

One corner case that would require handling is a race condition that happens when processing a dependency takes longer than the target artifact itself (as some steps have multiple workers, it might happen that the artifact outruns one of its dependencies). I would propose that (only) in such a case, the callable indexer should identify that we already have a result for a dependent (which then needs to be updated) and start a re-computation.

> what if some dependencies fail to process (and they often do), what will the callable indexer do then?

If one of the dependencies cannot be processed, we do not have partial results in the DB and the stitching has to happen on partial dependency information (same as we do right now). This is not nice, but we cannot do anything about it.

> why not incrementally update the graph DB of the callable indexer for each new package-version that it processes. I don't understand why that would not be possible.

Incrementally extending the index is not possible. Indexing calls requires to merge CGs and we would need to do this individually for *all dependents* of *every package* that is being added... this is quite an expensive operation and redoing so many times would be very wasteful.

> let's hope there are no cycles in the Maven dependency graph! Did anybody every check? Or whether our resolution algorithm can create dependency cycles.

I am not sure whether Maven even allows cycles, but even if, this is something that the recursive resolution strategy could take care of. Good catch.

Overall, I do not see a way around the proposed strategy, but we can keep thinking about it for a bit. If we cannot come up with a superior solution, I would advocate for creating a working solution first and observing the impact in practice, before we prematurely optimize for runtime at the cost of data consistency.

**Comment by proksch** [comment]
> > > 2- There will be repetitive computation for dependencies that are already processed. This can be solved by checking whether a package version is already inserted into the DB.
> > 
> > 
> > I think the REST API does that - if a package-version is already in the DB (package_versions table) or already requested for ingestion (ingested_artifacts table), it will not fire a new request.
> 
> Yes! The REST API does check whether the package version exists. If all agree on this solution, I can implement it in POMAnalyzer and create a PR.

Just to make sure that we talk about the same thing: yt is irrelevant, whether we have results in the DB, the question is whether it has been ingested already.

We certainly need to differentiate between dependencies that are drawn in from the pro-active Maven crawler (which should not congest the prio topics) and dependencies that are required by anything in the prio topics, which should have priority as well... as such, I do not think that the REST API is the best place, as it is not the information expert to decide these things:

1) It does not know about dependencies, it only knows about the requested artifact. I would not like to spread the same logic over two separate code locations (POMAnalyzer + REST API)
2) The resolution logic depends on the concrete language. There is no way that we will find a general handling, that can support Maven, PyPi and Debian at the same time.

For me, the POMAnalyzer (plus the equivalents in the other languages) seems be the obvious choice to handle this logic.

**Comment by proksch** [comment]
One other option would be to add a second input topic to the `POMAnalyzer` that acts as a working stack (LIFO, not sure if Kafka actually supports this).

When the POMAnalyzer requests a new task, there are basically two cases:

Case 1: The working stack is empty. We then consume one new artifact from the main topic. For the artifact, we check whether all dependencies have been ingested, if not, we add the missing dependencies and the artifact to the working stack.

Case 2: The working stack contains entries. The POMAnalyzer will grab the first artifact and perform the same logic.. if all dependencies have been ingested it can also ingest this one, otherwise, it will be put back on the stack, including all of its dependencies.

**Comment by mir-am** [comment]
> One other option would be to add a second input topic to the `POMAnalyzer` that acts as a working stack (LIFO, not sure if Kafka actually supports this).
> 
> When the POMAnalyzer requests a new task, there are basically two cases:
> 
> Case 1: The working stack is empty. We then consume one new artifact from the main topic. For the artifact, we check whether all dependencies have been ingested, if not, we add the missing dependencies and the artifact to the working stack.
> 
> Case 2: The working stack contains entries. The POMAnalyzer will grab the first artifact and perform the same logic.. if all dependencies have been ingested it can also ingest this one, otherwise, it will be put back on the stack, including all of its dependencies.
> 
> _edit:_ On second thought, also a FIFO queue would work, it just means that we no longer have the guarantee that artifacts are only added once...

A Kafka topic is basically a non-deterministic FIFO queue given that the topic has multiple partitions for parallelism. A one-partition topic is a classic FIFO queue.

**Comment by MagielBruntink** [comment]
Package-versions may be a dependency for various other package-versions. How will take factor in? Does the callable-indexer require only the dependencies, or also all dependents?

**Comment by MagielBruntink** [comment]
I just checked what the callable-index plugin actually does in the code. It only puts a GID graph, which is a graph using global callable IDs (created in the metadata DB on insertion) into RocksDB. The GID graphs appear to be package-version specific, there are no calls to dependencies in them, it seems. Can you point out to me where this merging process is actually happening in the code?

**Comment by proksch** [comment]
I was arguing based on a statement of @ashkboos, but maybe I was misunderstanding what he explained to me... I looked at the code myself and it seems you are right, there is no merging happening in the Indexer, just a reuse of the GID graph that has been generated by the metadata DB extension.

This would mean that we ONLY include the static call sites in the index, nothing more. These call sites *should* include calls to unresolved(!) external callables though, otherwise we cannot merge them afterwards anymore... It is a bit tricky to track what is being stored, as the data is being transported as a generic `JsonObject` in between the two plugins, so I first need to find a way to debug the plugin to say for sure.

I am wondering what this means for the search engine... without merging, the graphs do not make too much sense.

**Comment by mir-am** [comment]
As described in #347, Sebastiano is using the `CGMerger` API which merges a package version with all its dependency set.  The issue arises when we don't have CGs and GID Graphs for the whole (transitive) dependency set of a package version (e.g. `webgraph`).

I think for the search engine, @vigna is using `CGMerger` and he found issue #347.

**Comment by MagielBruntink** [comment]
Instead of high impact changes to the pipeline itself, I suggest we create a more advanced mvn crawler or a tool that requests a fully resolved dependency set to be ingested. This could run on demand, or even in continuous batch mode. But I would be against burdening POM analyzer and the default components with this task.

**Comment by mir-am** [comment]
@proksch
What do you think?
Should we extend POMAnalyzer to take care of ingesting resolved dep. set? or as Magiel suggested, we should create another plug-in or tool to do this?

**Comment by proksch** [comment]
I do not have a strong opinion on *where* we should add this functionality, but I do not follow the argumentation.

1) The POMAnalyzer *already identifies (direct) dependencies*, so no additional logic is required apart from adding these dependencies to the input topic. What is the high-impact change? Of course, we can split this out into another step in the pipeline, but I do not understand the advantage.

2) From my understanding it does not matter whether we extend the Maven crawler or the POMAnalyzer, as the result is the same... more than just one package will be added to the pipeline, which could lead to the "clogging" that @MagielBruntink has described before. We have agreed that the order of analysis seems to be less critical than originally thought, but I still see no way around the general idea of ingesting transitive closures -- unless we leave it to the client to make sure that the DB is consistent, which sounds like a terrible idea to me.

**Comment by MagielBruntink** [comment]
For ingestion of the individual package-versions involved in a transitive closure, the order of analysis of the package-versions does not matter. So a user / special tool would request all package-versions to be processed, after it is done, it can validate and then perform its work. In particular, the merging algorithms or the dependency graph generation can then be done. I don't see a big issue here.

**Comment by proksch** [comment]
What is the advantage of burdening the clients with this task compared to simply doing this in the POMAnalyzer, where this exact information is already available right now? I only see disadvantages with the proposed solution....

It seems that you have a use case in mind, for which the transitive closures seems to be irrelevant, which is why you prefer to have partial results quicker. I can acknowledge that... unfortunately, all other use cases (e.g., website, search-engine) need the transitive closure, so we would always need to trigger said handling and add the remaining packages... so regardless of how we implement it, the pipeline will have to do more work in the future.

Would it make sense to organize a brief call and talk about this issue? It seems that we are running in circles with the argumentation...

**Comment by proksch** [comment]
By the way: A very concrete problem is that a complex dependency set that is requested by the client (while being completely consistent) likely does not resolve to the same packages/versions that would be used when the different dependencies are being looked at in isolation... we need the latter 1) for the website and 2) for the generation of the exhaustive "universe" in the search engine that include all possible dependencies of all dependents of a particular package.

**Comment by MagielBruntink** [comment]
The advantage is simplicity. Although it depends on the change that is actually considered here. Some options:
 1. the POM Analyzer outputs information on each (transitive) dependency package-version on the downstream topic, ie. fasten.PomAnalyzer.out, in addition to the package-version that was anayzed.
 2. the POM Analyzer requests the REST API to trigger the ingestion of the (transitive) dependency package-versions, which then puts them on fasten.mvn.releases, there they are picked up by the POM Analyzer again.
 3. something more complicated, like explicit ordering or synchronization. In particular this scenario would be problematic in my view, as the pipeline is already complicated enough.

The benefit of scenario 2 is that the REST API already maintains an ingested_artifacts table for artifacts that are already in-flight. In scenario 1 potentially a lot of duplicate work can enter the pipeline.

**Comment by proksch** [comment]
The problem with Scenario 2 is that all dependencies then land on the priority track. I would vote for Scenario 1, with the extension that it also considers the ingestion table... we could actually move the check from the REST API to there, which seems to be the cleaner solution anyways:

Correct me, if my understanding is incorrect, but right now, packages that are added pro-actively to normal will be added a second time (to prio) when they are being requested before they have been processed... so they will reach the POMAnalyzer twice (or the CG generator or the metadata-DB-extension, not sure where the actual check is done that stops duplicate processing or whether this is done at all right now). A clean solution would check ingestion status in any way before starting processing and adding duplicates to the output topic.

A case that we need to discuss separately is what happens at the reset of the pipeline (e.g., when a plugin needs a bug fix that affects previous results), because this would create duplicates in subsequent pipeline steps by design unless we reset all topics and the ingestion table, but this goes beyond the discussion that we have right now...

**Comment by proksch** [comment]
On second thought... CG generation is much more expensive then POM parsing, so packages might have passed `POMAnalyzer` already when they get lazy-ingested into the prio track and they might wait for processing by `javacg-opal` or `metadata-db-extension` in the normal track. Maybe we just have to accept that there might be overlap between normal and prio... as the results should be deterministic, this does not present a big problem, as the only thing that we loose for duplicates is computation time. The first result will just be overwritten by a second, identical result.

An alternative, clean solution would stop processing in the current track, when it realizes that results already exist, but if we want to have a proper state handling and enable short-cuts in the process, we would need to track processing state for each particular package, but this goes both beyond what Kafka can offer and beyond what we want to implement ourselves.

**Comment by MagielBruntink** [comment]
> The problem with Scenario 2 is that all dependencies then land on the priority track. I would vote for Scenario 1, with the extension that it also considers the ingestion table... we could actually move the check from the REST API to there, which seems to be the cleaner solution anyways:

True, and yes, that is a good argument for Scenario 1.

> Correct me, if my understanding is incorrect, but right now, packages that are added pro-actively to normal will be added a second time (to prio) when they are being requested before they have been processed... so they will reach the POMAnalyzer twice (or the CG generator or the metadata-DB-extension, not sure where the actual check is done that stops duplicate processing or whether this is done at all right now). A clean solution would check ingestion status in any way before starting processing and adding duplicates to the output topic.

Also true. Nothing stops duplicate processing now, except for the REST API not doing a second ingestion request on a package-version that is not in the DB, but is in the ingested_artifacts table (so it is in flight somewhere in the pipeline).

> A case that we need to discuss separately is what happens at the reset of the pipeline (e.g., when a plugin needs a bug fix that affects previous results), because this would create duplicates in subsequent pipeline steps by design unless we reset all topics and the ingestion table, but this goes beyond the discussion that we have right now...

Yes. Right now the uncomplicated nature of the pipeline makes it easy to "re-do" a package-version. If we stop "re-doing" work early in the pipeline, which may be a good idea, we need to think about the maintenance scenario separetely. Perhaps yet another lane, ie. normal, priority, unconditional. Just throwing it out there ;-)

**Comment by MagielBruntink** [comment]
Well, given @proksch points above, I would also be in favor of Scenario 1 and not use the REST API for this. Instead, POM Analyzer would:
0. Initialize a set of package-versions to process
1. Receive a package-version (Maven coordinate) from its input topic, add to the set
2. Resolve its transitive dependencies, adding the resulting package-versions to the set
3. For each package-version in the set:
- check if it is currently in the DB already (package-versions table) OR in flight (in the ingested_artifacts table), if so, skip
- otherwise output the package-version and all related info to the POMAnalyzer output topic(s)

**Comment by mir-am** [comment]
Yes. It is possible to add a field to messages of `fasten.mvn.ingest` and check its value in POMAnalyzer.

**Comment by mir-am** [comment]
This thread has already a lot of posts and is hard to follow.
As suggested, I'd like to close this issue and discuss the proposed solution in #367.

**Comment by proksch** [comment]
I am sorry, but I disagree. The thread is very focussed on one particular problem so far.... splitting the discussion up into a second issue does not help with traceability.

**Comment by mir-am** [comment]
> Well, given @proksch points above, I would also be in favor of Scenario 1 and not use the REST API for this. Instead, POM Analyzer would:
> 
> 1. Initialize a set of package-versions to process
> 2. Receive a package-version (Maven coordinate) from its input topic, add to the set
> 3. Resolve its transitive dependencies, adding the resulting package-versions to the set
> 4. For each package-version in the set:
>    -- check if it is currently in the DB already (package-versions table) OR in flight (in the ingested_artifacts table), if so, skip
>    -- otherwise output the package-version and all related info to the POMAnalyzer output topic(s)
> 
> Edit: We could also consider adding a data field to the messages in the input topic of POMAnalyzer (fasten.mvn.releases) that flags whether the package-version should be processed even if is it already in the DB. That is one easy check in the POM Analyzer, right?

@MagielBruntink @proksch
I think that there is a downside to having a working set for POMAnlyzer. The plug-in keeps adding direct deps. to its working set for further processing. However, this set is in memory, i.e., if the plug-in crashes for whatever reason, the working set's members are gone. In fact, the plug-in does crash deliberately to release TCP connections (see #331 ).

Moreover, note that the current logic of the FASTEN server is: 1 Kafka message in (consume), 1 Kafka message out (produce).

ATM, I can think of two solutions to the described downside:
1- Lazy-ingesting discovered direct deps. to the priority topic. This delays the processing of a full (transitive) dep. set but eventually, they will be processed.
2- Creating another processing lane in the FASTEN server, which has precedence over the other two lanes (i.e. normal & priority). This guarantees that a full (transitive) dep. set will be processed together assuming that there will be no error further down the pipeline.

**Comment by proksch** [comment]
I am not sure about the third lane, but I think your concerns about persisting the intermediate steps are valid.

Maybe we just need another intermediate topic. So (on both the normal and the priority lane) we have an `input_deps` topic, which will be filled from REST API and from the Maven crawler with package versions of interest. A new plugin consumes these messages resolves the dependency set and sends one new message for each discovered dependency to an `expanded_deps` topic. This topic will then be consumed by the POMAnalyzer, which outputs its results as usual. This would be a simple solution that allows us to "add messages in the middle of a topic".

**Comment by MagielBruntink** [comment]
The resolution done for a package-version cannot be done for just direct dependencies. In Maven, for instance, there can be the dependencyManagement section that controls the versions of (transitive) dependencies. If you would ingest only the resolved direct dependencies, and they are later on further resolved, you don't have the dependencyManagement information available anymore and the versions may turn out different. 

You need to resolve once, and then request all transitive deps for that package-version to be processed.

You also need to record at that same moment that all those deps were requested for processing, so that deps are not processed over and over again.

**Comment by MagielBruntink** [comment]
With regards to crashing: if POM Analyzer crashes in the middle of processing a package-version, it's tough luck. It should signal the error to the error topic. I'm not sure why crashing is the best solution of managing the HTTP connection pool, surely there is a better approach possible?

I like the idea of a table that tracks package-versions through the pipeline, with a column for a number of milestones to reach, eg:
 - requested through REST API
 - resolved
 - call graph created
 - in MetaData DB
 - in callable index DB
 - vulnerabilities checked
 - quality metadata added
 - licensing metadata added

**Comment by proksch** [comment]
This should be solved through PR #372. Please re-open should further issues appear.
